Flask-like reloader for development

creatorrr opened the pull request:
Hey @szastupov, Been using aiotg in a project and it's been a charm! A flask-like reloader could be really helpful during development. I have put together a shabby implementation here. Do you think it is something you'd want to include in the module? If so, I can polish this PR further for a merge.

szastupov commented:
Hey @creatorrr ! The PR looks promising so I would love to include this feature, thanks!

creatorrr commented:
Sweet. Let me polish things up a bit and then push again.

szastupov commented on aiotg/bot.py:
```diff
@@ -100,6 +101,34 @@ def run(self):
         except KeyboardInterrupt:
             self.stop()
 
+    def run_with_reloader(self, **kwargs):
```
Maybe we should keep this function "private" but expose reloading through debug (or reload) option to run(), like flask does?

creatorrr replied:
Sure, or maybe add a `reload` flag to the main `run()` function. What do you prefer?

szastupov replied:
Yeah, let's add `reload` flag.

creatorrr replied:
Ok, pushing now.

creatorrr commented:
Done. Tested informally but I think a unit test for this would be too complicated. Although, I can work on it later if you think it is needed.

szastupov commented on .gitignore:
```diff
@@ -11,7 +12,7 @@ aiotg.egg-info
 # Must end with CRCR (^M^M in vim)
-Icon

+Icon
```
Is it intended?

creatorrr replied:
Nope. Probably my emacs settings

szastupov commented on aiotg/bot.py:
```diff
@@ -84,22 +85,37 @@ def no_handle(mt):
         """
         Convenience method for running bots in getUpdates mode
 
+        :param bool reload: Run with reloader
```
Something like "Automatically reload bot on code change" would sound better.

szastupov commented on aiotg/reloader.py:
```diff
@@ -0,0 +1,65 @@
+        )
+        os._exit(os.EX_OK)
+
+async def run_with_reloader( loop, coroutine, cleanup=lambda _:_, *args, **kwargs ):
```
`cleanup=lambda _:_` is clever but cleanup=None and explicit check before calling it would be a bit more cleaner and easier to follow ;)

creatorrr replied:
Hmm. Agreed, will make the change

szastupov commented on requirements.txt:
```diff
@@ -1 +1,11 @@
 aiohttp==1.0.0
+aionotify==0.2.0
```
You don't really have to touch this file (actually I'm gonna remove it, it's not used anyway).
Dependencies are set in install_requires section of setup.py, you cann add paco and aionotify there.

creatorrr replied:
ah, okay

creatorrr commented:
Fixed all the review request changes but Travis is still failing saying `ModuleNotFoundError: No module named 'aionotify'`.

@szastupov, I have added the deps to `setup.py`, do I need to do something else for them to work?

szastupov commented:
Oh, snap, I forgot that CI uses requirements-dev.text, sorry (can't wait for Pipfile to become stable so we don't have to deal with requirements files anymore). 

Just add modules to requirements-dev.txt without versions.

szastupov commented:
> UPDATE: Travis works now except throwing a weird The command "flake8" exited with 1 error.

Well, there are style errors :) It's not like I'm pep8 nazi but some style enforcement is better than nothing ;)

creatorrr commented:
True that. I can't see which errors exactly in the CI log, can you point them out or should I just run `flake8` on my branch?

creatorrr commented:
All passing now!

szastupov commented:
Looks good, nice job 👍

I was about to merge it but found out that aionotify is Linux-only, which is a major bummer since my primary dev machine is Macbook :(

creatorrr commented:
Oh crap! Let me look if there is an easy option to fix that.

creatorrr commented:
I think [asphalt-filewatcher](https://github.com/asphalt-framework/asphalt-filewatcher/blob/master/tests/test_watchers.py) could do the job. Has multiple backends for different platforms. But this is a significant rewrite, so I need some time (could also use some help ;) ).

szastupov commented:
asphalt-filewatcher looks a little bit too young. Have you checked [watchdog](http://pythonhosted.org/watchdog/)? That's what aiohttp-devtools is using.

Anyway, I'll see if I can help. We have nowhere to rush 👌

creatorrr commented:
Rewrote with watchdog. Should work like a charm now.

szastupov commented:
Whoah dude, that was fast 👍

szastupov commented on aiotg/bot.py:
```diff
@@ -84,22 +85,38 @@ def no_handle(mt):
         try:
-            loop.run_until_complete(self.loop())
+            if reload:
+                logger.setLevel( logging.DEBUG )
```
Maybe later we should introduce debug parameter to run() but for now it's good.

creatorrr replied:
Fair enough, I'll leave you to add that as you see fit.

szastupov commented on requirements-dev.txt:
```diff
@@ -1,5 +1,7 @@
 aiohttp
+aionotify
```
i guess now it's watchdog

creatorrr replied:
:+1:

szastupov commented on requirements.txt:
```diff
@@ -1 +1,9 @@
 aiohttp==1.0.0
+appdirs==1.4.3
```
No need to change this file.

creatorrr replied:
:+1:

szastupov commented:
Oh, one more thing! I guess we need to add some form of support for exclude lists because it gets triggered too often but maybe I'll just do it later myself, no worries.

creatorrr commented:
> Oh, one more thing! I guess we need to add some form of support for exclude lists because it gets triggered too often but maybe I'll just do it later myself, no worries.

You are right, the watcher was getting triggered way more often than intended. I was thinking of a graceful solution to this problem but for now, I have added the following whitelist to the watcher:

`["*.py", "*.txt", "*.aiml", "*.json", "*.cfg", "*.xml", "*.html"]`
(common filetypes I expect to be part of an aiotg project)

This should be good enough for most use cases but it's up to you how you'd want to make this more configurable. Let me know what you think

szastupov commented:
Great job, thank you 🙏
